src/eval.rs

```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
         Err(e) => report_error(&ecx, e),
     }
 }
+
+/// Turns an array of arguments into a Windows command line string.
+///
+/// The string will be UTF-16 encoded and NUL terminated.
+///
+/// Panics if the zeroth argument contains the `"` character because doublequotes
+/// in argv[0] cannot be encoded using the standard command line parsing rules.
+fn args_to_utf16_command_string<I, T>(mut args: I) -> Vec<u16>
+where
+    I: Iterator<Item = T>,
+    T: AsRef<str>,
+{
+    // Parse argv[0]. Slashes aren't escaped. Literal double quotes are not allowed.
+    let mut cmd = if let Some(arg0) = args.next() {
+        let arg0 = arg0.as_ref();
+        if arg0.is_empty() {
+            "\"\"".into()
+        } else if arg0.contains('"') {
+            panic!("argv[0] cannot contain a doublequote (\") character");
+        } else {
+            // Always surround argv[0] with quotes.
+            let mut s = String::new();
+            s.push('"');
+            s.push_str(arg0);
+            s.push('"');
+            s
+        }
+    } else {
+        return vec![0];
+    };
+
+    // Build the other arguments.
+    for arg in args {
+        let arg = arg.as_ref();
+        cmd.push(' ');
+        if arg.is_empty() {
+            cmd.push_str("\"\"");
+        } else if !arg.bytes().any(|c| matches!(c, b'"' | b'\t' | b' ')) {
+            cmd.push_str(arg);
+        } else {
+            cmd.push('"');
+            let mut chars = arg.chars().peekable();
+            loop {
+                let mut nslashes = 0;
+                while let Some(&'\\') = chars.peek() {
+                    chars.next();
+                    nslashes += 1;
+                }
+
+                match chars.next() {
+                    Some('"') => {
+                        cmd.extend(iter::repeat('\\').take(nslashes * 2 + 1));
+                        cmd.push('"');
+                    }
+                    Some(c) => {
```

Comment:
So the argument `\ ` becomes just `"\ "`, no further escaping for that backslash?

Reply:
Yes. From [Parsing C++ command-line arguments](https://docs.microsoft.com/en-us/cpp/cpp/main-function-command-line-args?view=msvc-160#parsing-c-command-line-arguments):

> The rule is that one or more `\` only need to be escaped if they are immediately followed by `"`.

I think I should add that link as a comment as well.

Reply:
Yes that is a helpful link.

From that text it is actually not clear to me that one can use `\"` inside double quote marks; the text starting "A string surrounded by double quote marks" only mentions "a pair of double quote marks" and does not say anything about `\` being special.  Also none of the examples cover this. So it is not clear to me from that text if `"a\\b"` will be interpreted as `a\\b` or `a\b` ending up in argv.

Reply:
Perhaps this is a more informative link: https://daviddeley.com/autohotkey/parameters/parameters.htm#WINCRULES

It goes through examples and defines the rules a bit more precisely.